src/power_grid_model/core/power_grid_model.py

```diff
@@ -149,92 +157,95 @@ def get_indexer(self, component_type: str, ids: np.ndarray):
         assert_no_error()
         return indexer
 
-    # pylint: disable=too-many-locals
-    # pylint: disable=too-many-branches
-    # pylint: disable=too-many-arguments
-    def _calculate(
+    def _output_type(self, options: Options):
+        return get_output_type(calculation_type=options.calculation_type, symmetric=options.symmetric)
+
+    def _construct_output(
         self,
+        output_component_types: Optional[Union[Set[str], List[str]]],
+        options: Options,
+        batch_size: int,
+    ) -> Dict[str, np.ndarray]:
+        # prepare result dataset
+        all_component_count = self.all_component_count
+
+        # for power flow, there is no need for sensor output
+        if options.calculation_type == CalculationType.power_flow:
+            all_component_count = {k: v for k, v in all_component_count.items() if "sensor" not in k}
+
+        # limit all component count to user specified component types in output
+        if output_component_types is None:
+            output_component_types = set(all_component_count.keys())
+
+        return create_output_data(
+            output_component_types=output_component_types,
+            output_type=self._output_type(options=options),
+            all_component_count=all_component_count,
+            batch_size=batch_size,
+        )
+
+    # pylint: disable=too-many-arguments
+    @staticmethod
+    def _options(
         calculation_type: CalculationType,
         symmetric: bool,
         error_tolerance: float,
         max_iterations: int,
         calculation_method: Union[CalculationMethod, str],
-        update_data: Optional[Dict[str, Union[np.ndarray, Dict[str, np.ndarray]]]],
         threading: int,
```

Comment:
some of the options are not relevant for short circuit. Maybe instead of list all the options, just use `**kw_options` in the argument?

Reply:
i thought about that beforehand, yes, but i did struggle with implementation. will retry 😄